### PR TITLE
Add PWM rainbow color gradient sample

### PR DESCRIPTION
**Pull request by decsny**
I made this sample because I find it more visually appealing then the existing rgb_led sample.

### CODE REVIEW

File: samples/basic/rgb_gradient/sample.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+sample:
+  name: RGB LED GRADIENT
```
Comment (henrikbrixandersen) [inline samples/basic/rgb_gradient/sample.yaml:2]:
```suggestion
  name: RGB LED Gradient
```

File: samples/basic/rgb_gradient/sample.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+tests:
+  sample.basic.rgb_gradient:
+    filter: dt_alias_exists("red-pwm-led") and
+            dt_alias_exists("green-pwm-led") and
+            dt_alias_exists("blue-pwm-led")
+    tags: drivers pwm
+    depends_on: pwm
+    harness: led
```
Comment (henrikbrixandersen) [inline samples/basic/rgb_gradient/sample.yaml:10]: No harness needed?

File: samples/basic/rgb_gradient/src/RGB_states.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+
+/* 0-255 based on color codes */
+typedef struct RGB_color {
```
Comment (henrikbrixandersen) [inline samples/basic/rgb_gradient/src/RGB_states.h:10]: Please avoid typedefs as per the coding style guidelines.

File: samples/basic/rgb_gradient/src/RGB_states.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	uint8_t red;
+	uint8_t green;
+	uint8_t blue;
+} Color;
+
+const Color None = {
```
Comment (henrikbrixandersen) [inline samples/basic/rgb_gradient/src/RGB_states.h:16]: Please avoid mixed-case variable names as per the coding style guidelines.

File: samples/basic/rgb_gradient/src/main.c [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	&Red, &Orange, &Yellow,
+	&Lime, &Green, &Aquamarine,
+	&Cyan, &Blue, &Purple, &Pink
+};
+
+const int num_colors = sizeof(colors);
+
+#define PWM_SLEEP_TIME K_MSEC(40)
+
+static void RGB_switch_state(Color *current_color, const Color *new_color)
```
Comment (henrikbrixandersen) [inline samples/basic/rgb_gradient/src/main.c:38]: Please avoid mixed-case function names.

File: samples/basic/rgb_gradient/src/main.c [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+	red_period = red_pwm_led.period;
+	green_period = green_pwm_led.period;
+	blue_period = blue_pwm_led.period;
+
+	red_color_unit_pulse_step = red_period / 255;
+	green_color_unit_pulse_step = green_period / 255;
+	blue_color_unit_pulse_step = blue_period / 255;
+
+	printf("Disco mode activated.\n");
```
Comment (henrikbrixandersen) [inline samples/basic/rgb_gradient/src/main.c:90]: 🕺

**Review by henrikbrixandersen — DISMISSED** [review]

File: samples/basic/rgb_gradient/sample.yaml (hunk 2 above)
Comment (decsny) [reply samples/basic/rgb_gradient/sample.yaml:10]: the other samples using LEDs have this harness, including rgb_led, blinky, blinky_pwm, and fade_led

File: samples/basic/rgb_gradient/src/RGB_states.h (hunk 3 above)
Comment (decsny) [reply samples/basic/rgb_gradient/src/RGB_states.h:10]: I guess this is not directly mentioned in the zephyr style guide but comes from the linux style guide which zephyr suggests to follow, so I'll change it

**Comment by decsny** [comment]
@henrikbrixandersen

**Review by kartben — CHANGES_REQUESTED** [review]

File: samples/basic/rgb_gradient/README.rst [hunk 7]
```diff
@@ -0,0 +1,56 @@
+.. _rgb-gradient-sample:
+
+PWM: RGB GRADIENT
+#################
```
[the remark this answers is not in the document]
Comment (decsny) [reply samples/basic/rgb_gradient/README.rst]: addressed

**Comment by henrikbrixandersen** [comment]
I have tested the sample on the `twr_ke18f`, not sure if it is intentional or not, but the color changes are very abrupt and not what I would call visually appealing.

**Comment by decsny** [comment]
@kartben please revisit

**Comment by kartben** [comment]
> I have tested the sample on the `twr_ke18f`, not sure if it is intentional or not, but the color changes are very abrupt and not what I would call visually appealing.

FWIW I am seeing the same (`az3166_iotdevkit` board) so I would tend to -1 until these glitches are addressed.
Not sure what the actual behavior is supposed to be, but in any case why not "fix" the initial rgb_led sample to make it smoother / more appealing?

File: samples/basic/rgb_gradient/src/main.c [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+uint32_t red_period;
+uint32_t green_period;
+uint32_t blue_period;
+
+int32_t red_color_unit_pulse_step;
+int32_t green_color_unit_pulse_step;
+int32_t blue_color_unit_pulse_step;
+
+const struct rgb_color *colors[10] = {
```
Comment (kartben) [inline samples/basic/rgb_gradient/src/main.c:28]: define the magic number

File: samples/basic/rgb_gradient/src/main.c [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+static void rgb_switch_state(struct rgb_color *current_color, const struct rgb_color *new_color)
+{
+	int32_t red_pulse_step = (new_color->red - current_color->red) *
+					red_color_unit_pulse_step / 10;
+	int32_t green_pulse_step = (new_color->green - current_color->green) *
+					green_color_unit_pulse_step / 10;
+	int32_t blue_pulse_step = (new_color->blue - current_color->blue) *
+					blue_color_unit_pulse_step / 10;
+
+	for (int i = 0; i < 10; i++) {
```
Comment (kartben) [inline samples/basic/rgb_gradient/src/main.c:47]: ditto re `10` magic number

**Comment by decsny** [comment]
> > I have tested the sample on the `twr_ke18f`, not sure if it is intentional or not, but the color changes are very abrupt and not what I would call visually appealing.
> 
> FWIW I am seeing the same (`az3166_iotdevkit` board) so I would tend to -1 until these glitches are addressed. Not sure what the actual behavior is supposed to be, but in any case why not "fix" the initial rgb_led sample to make it smoother / more appealing?

The original sample does something different than what this is doing, it's not a "fix", it's just a different pattern.

File: samples/basic/rgb_gradient/README.rst [hunk 10]
```diff
@@ -0,0 +1,56 @@
+.. zephyr:code-sample:: rgb-gradient
+
+RGB PWM Gradient
+################
```
Comment (kartben) [inline samples/basic/rgb_gradient/README.rst:4]: the directive calls for more actual metadata, and the name of the sample is redundant with the heading

```suggestion
.. zephyr:code-sample:: rgb-gradient
   :name: PWM RGB Gradient
   :relevant-api: pwm_interface

   Drive an RGB LED an RGB LED in a color gradient pattern using PWM.
```

**Review by kartben — CHANGES_REQUESTED** [review]
more feedback, independently from comments regarding this being potentially redundant with rgb_led and the fact it seems to have visual glitches

**Comment by decsny** [comment]
Okay, I did not think this would be so controversial, I don't care about this enough to continue for now

**Comment by kartben** [comment]
> > > I have tested the sample on the `twr_ke18f`, not sure if it is intentional or not, but the color changes are very abrupt and not what I would call visually appealing.
> > 
> > 
> > FWIW I am seeing the same (`az3166_iotdevkit` board) so I would tend to -1 until these glitches are addressed. Not sure what the actual behavior is supposed to be, but in any case why not "fix" the initial rgb_led sample to make it smoother / more appealing?
> 
> The other sample does something different than what this is doing, it's not a "fix", it's just a different pattern. I am not sure what are these glitches you guys see, I tried it on 4 or 5 different boards and it seemed fine to me

https://github.com/zephyrproject-rtos/zephyr/assets/128251/3eb30db6-0d1f-4260-be70-7c951d4d7313

**Comment by decsny** [comment]
> > > > I have tested the sample on the `twr_ke18f`, not sure if it is intentional or not, but the color changes are very abrupt and not what I would call visually appealing.
> > > 
> > > 
> > > FWIW I am seeing the same (`az3166_iotdevkit` board) so I would tend to -1 until these glitches are addressed. Not sure what the actual behavior is supposed to be, but in any case why not "fix" the initial rgb_led sample to make it smoother / more appealing?
> > 
> > 
> > The other sample does something different than what this is doing, it's not a "fix", it's just a different pattern. I am not sure what are these glitches you guys see, I tried it on 4 or 5 different boards and it seemed fine to me
> 
> PXL_20240611_143356709.mp4

Yes, I am not seeing these glitches on the boards I had, I don't know what the cause is

**Pull request closed without merging**